cedar-policy-validator/src/lib.rs

```diff
@@ -206,6 +286,192 @@ impl Validator {
         typecheck.typecheck_policy(t, &mut errors, &mut warnings);
         (errors.into_iter(), warnings.into_iter())
     }
+
+    #[cfg(feature = "level-validate")]
+    /// Check that `t` respects `max_allowed_level`
+    /// This assumes that (strict) typechecking has passed
+    fn check_entity_deref_level<'a>(
+        &'a self,
+        t: &'a Template,
+        max_allowed_level: &EntityDerefLevel,
+        policy_id: &PolicyID,
+    ) -> Vec<ValidationError> {
+        let typechecker = Typechecker::new(&self.schema, ValidationMode::Strict, t.id().clone());
+        let type_annotated_asts = typechecker.typecheck_by_request_env(t);
+        let mut errs = vec![];
+        for (_, policy_check) in type_annotated_asts {
+            match policy_check {
+                PolicyCheck::Success(e) => {
+                    let res =
+                        self.check_entity_deref_level_helper(&e, max_allowed_level, policy_id);
+                    match res.1 {
+                        Some(e) => errs.push(ValidationError::EntityDerefLevelViolation(e)),
+                        None => (),
+                    }
+                }
+                PolicyCheck::Fail(e) => println!("Typechecking failed: {:?}", e),
+                PolicyCheck::Irrelevant(_) => (), //Don't report level violations if typechecking already failed
```

Comment:
This might not be sound. IIRC, the irrelevant case includes policies with type `False`, but something like `principal.foo.bar.baz && false` has type `False` while still being a potential level violation.

Reply:
Fixed